File: Lab1/main.cpp
#include <iostream>
#include "Functions\Sinx_x3.h"
#include "Algorithms\DichotomyAlgorithm.h"
#include "Algorithms\GoldenRatioAlgorithm.h"
#include "Algorithms\FibonacciAlgorithm.h"
#include "Algorithms\ParabolaAlgorithm.h"
#include "Algorithms\BrentAlgorithm.h"

int main() {
    Sinx_x3 func;
    double leftBound, rightBound, eps;

    std::cout << "Enter left bound, right bound, precision (number of digits after decimal point)\n";
    std::cin >> leftBound >> rightBound >> eps;
    std::cout.precision(fabs(eps));

    DichotomyAlgorithm dichotomyAlgorithm;
    std::cout << "DichotomyAlgorithm:\n";
    std::cout << dichotomyAlgorithm.GetMin(func, leftBound, rightBound, pow(10, -eps)) << "\n";

    GoldenRationAlgorithm goldenRationAlgorithm;
    std::cout << "\nGoldenRationAlgorithm:\n";
    std::cout << goldenRationAlgorithm.GetMin(func, leftBound, rightBound, pow(10, -eps)) << "\n";

    FibonacciAlgorithm fibonacciAlgorithm;
    std::cout << "\nFibonacciAlgorithm:\n";
    std::cout << fibonacciAlgorithm.GetMin(func, leftBound, rightBound, pow(10, -eps)) << "\n";

    ParabolaAlgorithm parabolaAlgorithm;
    std::cout << "\nParabolaAlgorithm:\n";
    std::cout << parabolaAlgorithm.GetMin(func, leftBound, rightBound, pow(10, -eps)) << "\n";

    BrentAlgorithm brentAlgorithm;
    std::cout << "\nBrentAlgorithm:\n";
    std::cout << brentAlgorithm.GetMin(func, leftBound, rightBound, pow(10, -eps)) << "\n";

    return 0;
}

File: Lab1/main.py
import math
from algorithms import dichotomy_algorithm, golden_ratio_algorithm, \
    fibonacci_algorithm, parabola_algorithm, brent_algorithm


def our_function(x):
    return math.sin(x) * math.pow(x, 3)


def unimodal_function(x):
    return math.log(1 / x, 10) + pow(x, 0.7)


# print("Our function:")
#
# print("Dichotomy algorithm:")
# ans, calls, segments = dichotomy_algorithm.find_min(our_function, 3, 6, 1e-8)
# print(calls, len(segments), ans)
# print(segments, "\n")
#
# print("GoldenRatio algorithm:")
# ans, calls, segments = golden_ratio_algorithm.find_min(our_function, 3, 6, 1e-8)
# print(calls, len(segments), ans)
# print(segments, "\n")
#
# print("Fibonacci algorithm:")
# ans, calls, segments = fibonacci_algorithm.find_min(our_function, 3, 6, 1e-8)
# print(calls, len(segments), ans)
# print(segments, "\n")
#
# print("Parabola algorithm:")
# ans, calls, segments = parabola_algorithm.find_min(our_function, 3, 6, 1e-8)
# print(calls, len(segments), ans)
# print(segments, "\n")
#
# print("Brent algorithm:")
# ans, calls, segments = brent_algorithm.find_min(our_function, 3, 6, 1e-8)
# print(calls, len(segments), ans)
# print(segments, "\n")
#
#
# print("Unimodal function:")
#
# print("Dichotomy algorithm:")
# ans, calls, segments = dichotomy_algorithm.find_min(unimodal_function, 0.1, 2, 1e-8)
# print(calls, len(segments), ans)
# print(segments, "\n")
#
# print("GoldenRatio algorithm:")
# ans, calls, segments = golden_ratio_algorithm.find_min(unimodal_function, 0.1, 2, 1e-8)
# print(calls, len(segments), ans)
# print(segments, "\n")
#
# print("Fibonacci algorithm:")
# ans, calls, segments = fibonacci_algorithm.find_min(unimodal_function, 0.1, 2, 1e-8)
# print(calls, len(segments), ans)
# print(segments, "\n")
#
# print("Parabola algorithm:")
# ans, calls, segments = parabola_algorithm.find_min(unimodal_function, 0.1, 2, 1e-8)
# print(calls, len(segments), ans)
# print(segments, "\n")
#
# print("Brent algorithm:")
# ans, calls, segments = brent_algorithm.find_min(unimodal_function, 0.1, 2, 1e-8)
# print(calls, len(segments), ans)
# print(segments, "\n")


print("Dichotomy algorithm:")
ans, calls, segments = dichotomy_algorithm.find_min(our_function, 3, 6, 1e-4)
print(ans)

print("GoldenRatio algorithm:")
ans, calls, segments = golden_ratio_algorithm.find_min(our_function, -15, -3, 1e-4)
print(ans)

print("Fibonacci algorithm:")
ans, calls, segments = fibonacci_algorithm.find_min(our_function, -15, -3, 1e-4)
print(ans)

print("Parabola algorithm:")
ans, calls, segments = parabola_algorithm.find_min(our_function, -15, -3, 1e-4)
print(ans)

print("Brent algorithm:")
ans, calls, segments = brent_algorithm.find_min(our_function, -15, -3, 1e-4)
print(ans)
